src/megatron/energon/av/decode_av_frames.py

```diff
@@ -0,0 +1,441 @@
+                out_frame_size=video_out_frame_size,
+                decode_audio=video_decode_audio,
+                num_clips=audio_num_clips,
+                clip_duration=audio_clip_duration,
+            )
+        elif extension in ("flac", "mp3", "wav"):
+            media = decode_audio_samples(
+                self.stream,
+                num_clips=audio_num_clips,
+                clip_duration=audio_clip_duration,
+                audio_format=extension,
+            )
+        else:
+            return None
+
+        if media is not None:
+            frames = media[0].permute((0, 3, 1, 2)) if media[0] is not None else None
+            return VideoData(
+                frames=frames,
+                aframes=media[1],
+                info=media[2],
+            )
+        return None
+
+    def _get_extension(self) -> str | None:
+        """Get the file extension from the raw data."""
+        # Try to guess the file type using the first few bytes
+        self.stream.seek(0)  # Reset stream position before guessing
+        ftype = filetype.guess(self.stream)
+        if ftype is None:
+            return None
+        return ftype.extension
+
+def read_av_data(key: str, data: bytes) -> AVData:
+    """Decoder function that returns an AVData object for flexible decoding.
+
+    Args:
+        key: The file extension or key
+        data: The raw bytes of the media file
+
+    Returns:
+        AVData object that can be used to decode the media with custom parameters
+    """
+    return AVData(io.BytesIO(data))
+
+class AVDecoder:
+    def __init__(
+            self,
+            audio_clip_duration: int,
+            audio_num_clips: int,
+            video_decode_audio: bool,
+            video_num_frames: int,
+            video_out_frame_size: tuple[int, int],
+    ) -> None:
+        self.audio_clip_duration = audio_clip_duration
+        self.audio_num_clips = audio_num_clips
+        self.video_decode_audio = video_decode_audio
+        self.video_num_frames = video_num_frames
+        self.video_out_frame_size = video_out_frame_size
+
+    def __call__(self, key: str, data: bytes) -> VideoData | None:
+        """
+        Extract the video or audio data from default media extensions.
+
+        Args:
+            key: media file extension
+            data: raw media bytes
+
+        Returns:
+            VideoData containing the decoded frames and metadata, or None if decoding failed
+        """
+        av_data = read_av_data(key, data)
+        if av_data is None:
+            return None
+        return av_data.get_frames(
+            audio_clip_duration=self.audio_clip_duration,
+            audio_num_clips=self.audio_num_clips,
+            video_decode_audio=self.video_decode_audio,
+            video_num_frames=self.video_num_frames,
+            video_out_frame_size=self.video_out_frame_size,
+        )
+
+
+def frame_to_ts(frame: int, average_rate: Fraction, time_base: Fraction) -> int:
```

Comment:
I'd suggest to move all these top-level functions (i.e. all below) to `AVData` class

Reply:
done